File: satisfiability/satisfiability.py
#python3
#Arturas Aleksandrauskas - artal938 TDDD95
import sys
from random import randint
from collections import defaultdict
import operator
def solve2(expresions, comb):
    t = []
    for e in expresions:
        satisfide = False
        for i, b in e:
            if b == "~":
                satisfide = satisfide or (comb ^ -1) & (1<<i)
            else:
                satisfide = satisfide or comb & (1<<i)

        if not satisfide:
            for i, b in e:
                t.append(i)
            break

    if len(t) > 0:
        #print(t)
        #e = t[randint(0,len(t)-1)]
        #return comb ^ 1 << e[randint(0,len(e)-1)][0]
        return comb ^ 1 << t[randint(0,len(t)-1)]
    return -1

def solve(expresions, comb, n):
    for i in range(2000):
        comb = solve2(expresions, comb)
        #print(comb)
        if comb == -1:
            return True
        if i%150 == 0:
            comb = randint(0,2**n)
    return False


def main():
    tests = sys.stdin.readline()
    for _ in range(int(tests)):
        n, m = map(int, sys.stdin.readline().split())
        expresions = [[]]*m
        for i in range(m):
            expresion = sys.stdin.readline().strip().split(' v ')
            expresion = [(int( i[2:]if i[0] == "~"else i[1:] )-1, i[0]) for i in expresion]
            expresions[i]= expresion
        expresions.sort(key=len)
        if solve(expresions, randint(0,2**n), n):
            print("satisfiable")
        else:
            print("unsatisfiable")
            
if __name__ == "__main__":
    main()

File: satisfiability/satisfiability.cpp
//Arturas Aleksandrauskas - artal938 TDDD95


#include <bits/stdc++.h>
using namespace std;

bool solve(vector<vector<pair<int, bool>>> & expresions, int comb){
    for(auto expresion: expresions){
        bool satisfide{false};
        for(auto e: expresion){
            if(e.second){
                satisfide = satisfide || ((~comb) & (1 << e.first));
            }else{
                satisfide = satisfide || (comb & (1 << e.first));
            }
        }
        if (!satisfide){
            return false;
        }
    }
    return true;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    int n{}, m{}, t{};
    string imput;
    cin >> t;
    for (int i{}; i < t; i++){
        cin >> n >> m;
        vector<vector<pair<int, bool>>> expresions;
        expresions.resize(m, vector<pair<int, bool>>());
        getline(cin,imput);
        for (int j{}; j < m; j++){
            getline(cin,imput); 
            bool neg{false};
            int pos{};
            for(char c: imput){
                if (c == ' '){
                    continue;
                }else if(c == '~'){
                    neg = true;
                }else if(c == 'X'){
                    continue;
                }else if(c == 'v'){
                    expresions[j].push_back({pos-1, neg});
                    pos = 0;
                    neg = false;
                }else{
                    pos *= 10;
                    pos += (int)c-48;
                }
            }
            expresions[j].push_back({pos-1, neg});
        }
        // cout << "---------------" << endl;
        // for (auto expre :expresions){
        //     for (auto e :expre){
        //         if(e.second){
        //             cout << "~" << e.first  << " ";
        //         }else{
        //             cout << e.first << " ";
        //         }
        //     }
        //     cout << endl;
        // }
        bool sati{false};
        for(int j{(1<<n)-1}; j >= 0; j--){
            if (solve(expresions, j)){
                cout << "satisfiable" << endl;
                sati = true;
                break;
            }
        }
        if(!sati){
            cout << "unsatisfiable" << endl;
        }
    }

    
    return 0;
}